File: frontend/public/sdk/patchflow.js
/**
 * PatchFlow Agent SDK — Node.js & Edge Runtime
 * ============================================
 *
 * Universal one-line error monitoring for Next.js (App Router, Pages Router,
 * Middleware, Edge Runtime), Express, Hono, NestJS, or any JavaScript / Node app.
 *
 * Usage
 * -----
 * Next.js (instrumentation.ts):
 *   import patchflow from './patchflow';
 *   patchflow.init({ apiKey: process.env.PATCHFLOW_API_KEY });
 *
 * Next.js Route Handler (route.ts):
 *   export const GET = patchflow.wrapNextHandler(async () => { ... });
 *
 * Express:
 *   const patchflow = require('./patchflow');
 *   patchflow.init({ apiKey: process.env.PATCHFLOW_API_KEY });
 *   app.use(patchflow.expressMiddleware());
 *
 * Manual:
 *   try { riskyCode() } catch (e) { await patchflow.captureException(e) }
 */

'use strict';

const SDK_VERSION = '0.1.0';
const DEFAULT_HOST = 'https://patchflow-backend-xax6.onrender.com';

// ── Singleton ─────────────────────────────────────────────────────────────────

let _instance = null;

/**
 * Initialise the PatchFlow SDK.
 *
 * @param {object} options
 * @param {string}  options.apiKey        - Your site API key (pf_live_...)
 * @param {string}  [options.host]        - PatchFlow API host
 * @param {string}  [options.environment] - Environment name (default: 'production')
 * @param {boolean} [options.debug]       - Print debug logs
 * @returns {PatchFlow}
 */
function init({ apiKey, host, environment, debug = false } = {}) {
  const envHost = typeof process !== 'undefined' && process.env ? process.env.PATCHFLOW_HOST : undefined;
  const envName = typeof process !== 'undefined' && process.env
    ? (process.env.PATCHFLOW_ENV || process.env.NODE_ENV)
    : undefined;

  _instance = new PatchFlow({
    apiKey,
    host: host || envHost || DEFAULT_HOST,
    environment: environment || envName || 'production',
    debug,
  });

  // Install process-level handlers only if running in Node.js environment
  _installProcessHandlers(_instance);

  // Automatically send non-blocking startup heartbeat to mark SDK as active in dashboard
  _instance.ping();

  if (debug) {
    console.log(`[PatchFlow] Initialised. host=${_instance.host} env=${_instance.environment}`);
  }

  return _instance;
}

// ── Universal HTTP Dispatcher (Fetch-based, 100% Webpack & Edge Safe) ─────────

async function _dispatchHttp(urlStr, headers, bodyStr, timeoutMs = 10000) {
  try {
    const controller = typeof AbortController !== 'undefined' ? new AbortController() : null;
    const timer = controller ? setTimeout(() => controller.abort(), timeoutMs) : null;

    const res = await fetch(urlStr, {
      method: 'POST',
      headers: headers || {},
      body: bodyStr || undefined,
      signal: controller ? controller.signal : undefined,
    });
    if (timer) clearTimeout(timer);
    return res;
  } catch (_) {
    return null;
  }
}

// ── Core class ────────────────────────────────────────────────────────────────

class PatchFlow {
  constructor({ apiKey, host, environment, debug }) {
    this.apiKey = apiKey || '';
    this.host = (host || DEFAULT_HOST).replace(/\/$/, '');
    this.environment = environment || 'production';
    this.debug = Boolean(debug);
  }

  /**
   * Send a heartbeat ping to confirm connection and mark the SDK as active.
   */
  async ping() {
    if (!this.apiKey) return;
    try {
      const res = await _dispatchHttp(
        `${this.host}/api/sdk/ping`,
        {
          'X-PatchFlow-Key': this.apiKey,
          'User-Agent': `patchflow-node/${SDK_VERSION}`,
        },
        null,
        5000
      );
      if (this.debug) {
        console.log('[PatchFlow] Heartbeat ping dispatched:', res?.status || 'ok');
      }
      return res;
    } catch (e) {
      if (this.debug) console.warn('[PatchFlow] Failed to send heartbeat ping:', e);
    }
  }

  /**
   * Capture and send an exception to PatchFlow.
   *
   * @param {Error|any} error
   * @param {object}    [context]
   * @param {string}    [context.endpoint]   - Route path e.g. '/api/users'
   * @param {string}    [context.method]     - HTTP method
   * @param {number}    [context.statusCode] - HTTP status code
   * @param {string}    [context.framework]  - Framework name
   */
  async captureException(error, context = {}) {
    if (!error) return;
    try {
      const payload = buildPayload(error, {
        ...context,
        environment: this.environment,
      });
      return await this._send(payload);
    } catch (e) {
      if (this.debug) console.error('[PatchFlow] Failed to capture exception:', e);
    }
  }

  async _send(payload) {
    if (!this.apiKey) return;
    try {
      const body = JSON.stringify(payload);
      const res = await _dispatchHttp(
        `${this.host}/api/sdk/errors`,
        {
          'Content-Type': 'application/json',
          'X-PatchFlow-Key': this.apiKey,
          'User-Agent': `patchflow-node/${SDK_VERSION}`,
        },
        body,
        10000
      );
      if (this.debug) {
        console.log('[PatchFlow] Error payload dispatched:', res?.status || 'ok');
      }
      return res;
    } catch (e) {
      if (this.debug) console.error('[PatchFlow] Send failed:', e);
    }
  }
}

// ── Payload builder ───────────────────────────────────────────────────────────

function buildPayload(error, context = {}) {
  const errObj = error instanceof Error ? error : new Error(String(error));
  const frames = parseStack(errObj.stack || '');
  const top = frames[frames.length - 1] || {};
  const culprit = context.endpoint || `${top.filename || 'unknown'}:${top.lineno || 0}`;

  return {
    error_type: errObj.name || errObj.constructor?.name || 'Error',
    error_message: (errObj.message || String(errObj)).slice(0, 1000),
    culprit,
    endpoint: context.endpoint || '',
    method: (context.method || '').toUpperCase(),
    status_code: context.statusCode || null,
    stack_frames: frames,
    framework: context.framework || detectFramework(),
    environment: context.environment || 'production',
    sdk_version: SDK_VERSION,
  };
}

function parseStack(stack) {
  if (!stack) return [];
  const lines = stack.split('\n').slice(1);
  const frames = [];
  const atRegex = /^\s*at\s+(?:(.+?)\s+\()?(.+?):(\d+):(\d+)\)?/;

  for (const line of lines) {
    const m = atRegex.exec(line);
    if (m) {
      const [, fn, file, lineno] = m;
      if (file && !file.startsWith('node:') && !file.includes('patchflow.js')) {
        frames.push({
          filename: file,
          lineno: parseInt(lineno, 10) || 0,
          function: fn || '<anonymous>',
          context_line: '',
          pre_context: [],
          post_context: [],
          vars: {},
        });
      }
    }
  }
  return frames;
}

function detectFramework() {
  if (typeof process !== 'undefined' && process.env) {
    if (process.env.NEXT_RUNTIME || process.env.__NEXT_PROCESSED_ENV) return 'nextjs';
  }
  if (typeof globalThis !== 'undefined' && globalThis.EdgeRuntime) return 'nextjs-edge';
  return 'node';
}

// ── Express middleware ────────────────────────────────────────────────────────

/**
 * Express error-handling middleware.
 * Must be registered AFTER all routes: `app.use(patchflow.expressMiddleware())`
 *
 * @returns {(err: any, req: any, res: any, next: any) => void}
 */
function expressMiddleware() {
  const pf = _requireInstance('expressMiddleware');
  return function patchflowExpressMiddleware(err, req, res, next) {
    try {
      pf.captureException(err, {
        endpoint: req.originalUrl || req.url,
        method: req.method,
        statusCode: res.statusCode >= 400 ? res.statusCode : 500,
        framework: 'express',
      });
    } catch (_) {}
    next(err);
  };
}

// ── Next.js App Router wrapper ────────────────────────────────────────────────

/**
 * Wrap a Next.js App Router handler to capture unhandled errors on Serverless.
 *
 * Usage:
 *   export const GET = patchflow.wrapNextHandler(async (request) => {
 *     return NextResponse.json({ ok: true });
 *   });
 *
 * @template {(...args: any[]) => any} T
 * @param {T} handler
 * @returns {T}
 */
function wrapNextHandler(handler) {
  const pf = _requireInstance('wrapNextHandler');
  /** @type {any} */
  const wrapped = async function (request, context) {
    try {
      return await handler(request, context);
    } catch (err) {
      try {
        const urlStr = request && request.url ? request.url : '';
        const endpoint = urlStr ? new URL(urlStr).pathname : '/crash';
        await pf.captureException(err, {
          endpoint,
          method: request?.method || 'GET',
          statusCode: 500,
          framework: 'nextjs',
        });
      } catch (_) {}
      throw err;
    }
  };
  return /** @type {T} */ (wrapped);
}

// ── Hono middleware ───────────────────────────────────────────────────────────

/**
 * Hono middleware factory.
 * Usage: app.use('*', patchflow.honoMiddleware())
 *
 * @returns {(c: any, next: () => Promise<void>) => Promise<void>}
 */
function honoMiddleware() {
  const pf = _requireInstance('honoMiddleware');
  return async function patchflowHonoMiddleware(c, next) {
    try {
      await next();
    } catch (err) {
      try {
        await pf.captureException(err, {
          endpoint: new URL(c.req.url).pathname,
          method: c.req.method,
          statusCode: 500,
          framework: 'hono',
        });
      } catch (_) {}
      throw err;
    }
  };
}

// ── Process-level fallback (Node.js only) ──────────────────────────────────────

function _installProcessHandlers(pf) {
  if (typeof process === 'undefined' || typeof process.on !== 'function') {
    return; // Skip on Edge Runtime, Web Workers, or browsers
  }
  try {
    process.on('uncaughtException', (err) => {
      pf.captureException(err);
    });

    process.on('unhandledRejection', (reason) => {
      const err = reason instanceof Error ? reason : new Error(String(reason));
      pf.captureException(err);
    });
  } catch (_) {}
}

// ── Helpers ───────────────────────────────────────────────────────────────────

function _requireInstance(fnName) {
  if (!_instance) {
    throw new Error(
      `[PatchFlow] Call patchflow.init({ apiKey: '...' }) before using ${fnName}()`
    );
  }
  return _instance;
}

/**
 * Capture an exception using the globally initialised SDK instance.
 * @param {Error|any} error
 * @param {object} [context]
 */
async function captureException(error, context = {}) {
  if (_instance) return await _instance.captureException(error, context);
}

// ── Exports ───────────────────────────────────────────────────────────────────

module.exports = {
  init,
  captureException,
  expressMiddleware,
  wrapNextHandler,
  honoMiddleware,
  PatchFlow,
};

// ESM default export compatibility
module.exports.default = module.exports;

File: frontend/public/sdk/patchflow.py
"""
PatchFlow Agent SDK — Python
============================

One-line setup for FastAPI, Flask, Starlette, or any WSGI/ASGI framework.

Usage
-----
FastAPI / Starlette:
    import patchflow
    patchflow.init(api_key="pf_live_...")
    # That's it — middleware is installed automatically.

Flask:
    import patchflow
    patchflow.init(api_key="pf_live_...", app=flask_app)

Manual (any Python app):
    import patchflow
    pf = patchflow.init(api_key="pf_live_...")
    try:
        risky_code()
    except Exception as e:
        pf.capture_exception(e)

What it captures
----------------
- Unhandled exceptions in FastAPI / Starlette routes (via ASGI middleware)
- Unhandled exceptions in Flask routes (via Flask error handler)
- Any exception passed to capture_exception()

What it sends to PatchFlow
--------------------------
- Exception type and message (PII auto-redacted server-side)
- Stack frames (file, line, function, context lines)
- Endpoint path and HTTP method
- Framework name and SDK version
- Environment name

Privacy
-------
All data is sent over HTTPS. PatchFlow redacts emails, IPs, passwords,
tokens, and connection strings from error messages and stack frames
before storing or processing them.
"""

from __future__ import annotations

import inspect
import os
import sys
import threading
import traceback
from typing import Callable, Any

try:
    import httpx as _httpx
    _HTTP = "httpx"
except ImportError:
    import urllib.request as _urllib
    import json as _json
    _HTTP = "urllib"

__version__ = "0.1.0"
_DEFAULT_HOST = "https://patchflow-backend-xax6.onrender.com"

# ── Singleton ─────────────────────────────────────────────────────────────────

_instance: "PatchFlow | None" = None


def init(
    api_key: str,
    host: str | None = None,
    environment: str | None = None,
    app: Any = None,           # Flask app, optional
    debug: bool = False,
) -> "PatchFlow":
    """
    Initialise the PatchFlow SDK.

    Parameters
    ----------
    api_key:     Your site API key from the PatchFlow dashboard (pf_live_...)
    host:        PatchFlow API host (defaults to https://api.patchflow.dev)
    environment: Environment name (defaults to PATCHFLOW_ENV or APP_ENV env vars, or 'production')
    app:         Flask application instance — pass this for Flask projects
    debug:       Print SDK debug logs to stdout
    """
    global _instance
    _instance = PatchFlow(
        api_key=api_key,
        host=host or os.getenv("PATCHFLOW_HOST", _DEFAULT_HOST),
        environment=environment or os.getenv("PATCHFLOW_ENV") or os.getenv("APP_ENV", "production"),
        debug=debug,
    )

    # ── Auto-install Flask error handler ─────────────────────────────────────
    if app is not None:
        _install_flask(app, _instance)
        if debug:
            print(f"[PatchFlow] Flask error handler installed on {app}")

    # ── Auto-install FastAPI / Starlette ASGI middleware ──────────────────────
    # Detect if we're inside a FastAPI app by checking the call stack for
    # a FastAPI or Starlette Application object.
    else:
        _try_install_fastapi(_instance, debug)

    # ── Fallback: install sys.excepthook for non-web scripts ─────────────────
    _install_excepthook(_instance)

    if debug:
        print(f"[PatchFlow] Initialised. host={_instance.host} env={_instance.environment}")

    return _instance


# ── Core class ────────────────────────────────────────────────────────────────

class PatchFlow:
    def __init__(self, api_key: str, host: str, environment: str, debug: bool):
        self.api_key = api_key
        self.host = host.rstrip("/")
        self.environment = environment
        self.debug = debug
        self._framework = _detect_framework()

    def capture_exception(
        self,
        exc: BaseException,
        endpoint: str = "",
        method: str = "",
        status_code: int | None = None,
    ) -> None:
        """
        Capture and send an exception to PatchFlow.
        This is non-blocking — the HTTP call is made in a background thread.
        """
        try:
            payload = _build_payload(
                exc=exc,
                endpoint=endpoint,
                method=method,
                status_code=status_code,
                framework=self._framework,
                environment=self.environment,
            )
            threading.Thread(
                target=self._send,
                args=(payload,),
                daemon=True,
            ).start()
        except Exception as send_err:
            if self.debug:
                print(f"[PatchFlow] Failed to capture exception: {send_err}")

    def _send(self, payload: dict) -> None:
        url = f"{self.host}/api/sdk/errors"
        headers = {
            "X-PatchFlow-Key": self.api_key,
            "Content-Type": "application/json",
            "User-Agent": f"patchflow-python/{__version__}",
        }
        try:
            if _HTTP == "httpx":
                with _httpx.Client(timeout=10) as client:
                    r = client.post(url, json=payload, headers=headers)
                    if self.debug:
                        print(f"[PatchFlow] Sent error — status={r.status_code}")
            else:
                import json
                data = json.dumps(payload).encode()
                req = _urllib.Request(url, data=data, headers=headers, method="POST")
                with _urllib.urlopen(req, timeout=10) as r:
                    if self.debug:
                        print(f"[PatchFlow] Sent error — status={r.status}")
        except Exception as e:
            if self.debug:
                print(f"[PatchFlow] Send failed: {e}")


# ── Payload builder ───────────────────────────────────────────────────────────

def _build_payload(
    exc: BaseException,
    endpoint: str,
    method: str,
    status_code: int | None,
    framework: str,
    environment: str,
) -> dict:
    tb = exc.__traceback__
    frames = []
    if tb:
        extracted = traceback.extract_tb(tb)
        for frame in extracted:
            frames.append({
                "filename": frame.filename,
                "lineno": frame.lineno,
                "function": frame.name,
                "context_line": frame.line or "",
                "pre_context": [],
                "post_context": [],
                "vars": {},
            })

    # Top frame for culprit
    top = frames[-1] if frames else {}
    culprit = endpoint or f"{top.get('filename','')}:{top.get('lineno','')}"

    return {
        "error_type": type(exc).__name__,
        "error_message": str(exc)[:1000],
        "culprit": culprit,
        "endpoint": endpoint,
        "method": method.upper() if method else "",
        "status_code": status_code,
        "stack_frames": frames,
        "framework": framework,
        "environment": environment,
        "sdk_version": __version__,
    }


# ── Framework detection ───────────────────────────────────────────────────────

def _detect_framework() -> str:
    if "fastapi" in sys.modules:
        return "fastapi"
    if "flask" in sys.modules:
        return "flask"
    if "starlette" in sys.modules:
        return "starlette"
    if "django" in sys.modules:
        return "django"
    return "python"


# ── FastAPI / Starlette ASGI middleware ───────────────────────────────────────

def _try_install_fastapi(pf: "PatchFlow", debug: bool) -> None:
    """
    Look for a FastAPI or Starlette app in the calling module's globals
    and install the middleware automatically.
    """
    try:
        # Walk up the call stack looking for a FastAPI/Starlette app
        frame = inspect.currentframe()
        while frame:
            for name, obj in frame.f_globals.items():
                if _is_asgi_app(obj):
                    obj.add_middleware(PatchFlowASGIMiddleware, patchflow=pf)
                    if debug:
                        print(f"[PatchFlow] ASGI middleware installed on {name}")
                    return
            frame = frame.f_back
    except Exception:
        pass  # Silently skip — user can install middleware manually


def _is_asgi_app(obj: Any) -> bool:
    try:
        from starlette.applications import Starlette
        if isinstance(obj, Starlette):
            return True
    except ImportError:
        pass
    try:
        from fastapi import FastAPI
        if isinstance(obj, FastAPI):
            return True
    except ImportError:
        pass
    return False


class PatchFlowASGIMiddleware:
    """ASGI middleware that captures unhandled exceptions from FastAPI/Starlette."""

    def __init__(self, app: Any, patchflow: "PatchFlow"):
        self.app = app
        self.pf = patchflow

    async def __call__(self, scope: dict, receive: Any, send: Any) -> None:
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        endpoint = scope.get("path", "")
        method = scope.get("method", "")

        try:
            await self.app(scope, receive, send)
        except Exception as exc:
            self.pf.capture_exception(exc, endpoint=endpoint, method=method, status_code=500)
            raise  # re-raise so FastAPI's own error handling still runs


# ── Flask integration ─────────────────────────────────────────────────────────

def _install_flask(app: Any, pf: "PatchFlow") -> None:
    """Register a Flask after_request + errorhandler hook."""
    try:
        @app.errorhandler(Exception)
        def _pf_flask_error(exc: Exception):
            # Get request context
            try:
                from flask import request as flask_req
                endpoint = flask_req.path
                method = flask_req.method
            except Exception:
                endpoint, method = "", ""
            pf.capture_exception(exc, endpoint=endpoint, method=method, status_code=500)
            # Re-raise so Flask's own error handling continues
            raise exc
    except Exception:
        pass


# ── sys.excepthook fallback ───────────────────────────────────────────────────

def _install_excepthook(pf: "PatchFlow") -> None:
    """Capture unhandled exceptions in non-web scripts."""
    original = sys.excepthook

    def _hook(exc_type, exc_value, exc_tb):
        if issubclass(exc_type, KeyboardInterrupt):
            original(exc_type, exc_value, exc_tb)
            return
        pf.capture_exception(exc_value)
        original(exc_type, exc_value, exc_tb)

    sys.excepthook = _hook


# ── Convenience top-level functions ───────────────────────────────────────────

def capture_exception(exc: BaseException, **kwargs) -> None:
    """Capture an exception using the globally initialised SDK instance."""
    if _instance:
        _instance.capture_exception(exc, **kwargs)


# ── FastAPI convenience decorator ─────────────────────────────────────────────

def monitor(func: Callable) -> Callable:
    """
    Decorator for FastAPI route functions.
    Catches and reports exceptions without swallowing them.

    Usage:
        @app.get("/users")
        @patchflow.monitor
        async def get_users():
            ...
    """
    import functools

    @functools.wraps(func)
    async def _async_wrapper(*args, **kwargs):
        try:
            return await func(*args, **kwargs)
        except Exception as exc:
            if _instance:
                _instance.capture_exception(exc)
            raise

    @functools.wraps(func)
    def _sync_wrapper(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except Exception as exc:
            if _instance:
                _instance.capture_exception(exc)
            raise

    import asyncio
    if asyncio.iscoroutinefunction(func):
        return _async_wrapper
    return _sync_wrapper
